packages/react-native-renderer/src/ReactNativeGetListener.js

```diff
@@ -7,30 +7,82 @@
  */
 
 import type {Fiber} from 'react-reconciler/src/ReactInternalTypes';
+import type {PropagationPhases} from './legacy-events/PropagationPhases';
 
 import {getFiberCurrentPropsFromNode} from './legacy-events/EventPluginUtils';
 
 export default function getListener(
   inst: Fiber,
   registrationName: string,
+  phase: PropagationPhases,
 ): Function | null {
+  // Previously, there was only one possible listener for an event:
+  // the onEventName property in props.
+  // Now, it is also possible to have N listeners
+  // for a specific event on a node. Thus, we accumulate all of the listeners,
+  // including the props listener, and return a function that calls them all in
+  // order, starting with the handler prop and then the listeners in order.
+  // We still return a single function or null.
+  const listeners = [];
+
   const stateNode = inst.stateNode;
-  if (stateNode === null) {
-    // Work in progress (ex: onload events in incremental mode).
-    return null;
+  // If null: Work in progress (ex: onload events in incremental mode).
+  if (stateNode !== null) {
+    const props = getFiberCurrentPropsFromNode(stateNode);
+    if (props === null) {
+      // Work in progress.
+      return null;
+    }
+    const listener = props[registrationName];
+
+    if (listener && typeof listener !== 'function') {
+      throw new Error(
+        `Expected \`${registrationName}\` listener to be a function, instead got a value of \`${typeof listener}\` type.`,
+      );
+    }
+
+    if (listener) {
+      listeners.push(listener);
+    }
+  }
+
+  // Get imperative event listeners for this event
+  if (stateNode.canonical._eventListeners[registrationName] && stateNode.canonical._eventListeners[registrationName].length > 0) {
+    var toRemove = [];
+    for (var listenerObj of stateNode.canonical._eventListeners[registrationName]) {
+      // Make sure phase of listener matches requested phase
+      const isCaptureEvent = listenerObj.options.capture != null && listenerObj.options.capture;
+      if (isCaptureEvent !== (phase === 'captured')) {
+        continue;
+      }
+
+      listeners.push(listenerObj.listener);
+
+      // Only call once? If so, remove from listeners after calling.
+      if (listenerObj.options.once) {
+        toRemove.push(listenerObj);
+      }
+    }
+    for (var listenerObj of toRemove) {
+      stateNode.canonical.removeEventListener_unstable(registrationName, listenerObj.listener, listenerObj.capture);
```

Comment:
I don't have the context on the background of the APIs being added in this PR. But one thing that jumps out to me is that it seems like a hazard that calling `getListener` has a side effect of modifying the listeners. This seems very easy to accidentally mess up from the caller side because it looks like a pure function.

Reply:
Do you have a suggestion here? I agree - but currently with the way this API is used, these listeners *are* guaranteed to be called in the same frame; and if the semantic is "once", then we need to remove the listener or somehow mark it as "used". Keeping it around forever with a dirty flag doesn't seem good either for memory reasons. 

The "once" flag is a (w3c) standard flag, so it's necessary and useful to make sure it works properly. 

Thanks for the feedback.

Reply:
I think my instinct would be to wrap the listeners that have `{ once: true }` with code that removes the listener imperatively when it actually fires. Then the special case is rare and the `getListener` function stays pure. (Semantic question: if propagation was prevented by a handler above, does the parent handler with `{once: true}` get removed or not? Seems like this would affect how we implement it. `getListener` getting called does not guarantee the listener itself will get called.)

Reply:
I think this shows that in the DOM, the `{ once: true }` parent listener does not get removed if the child stopped propagation: https://codesandbox.io/s/serene-minsky-380ki?file=/src/index.js

This means that `getListener()` is too early to remove the `once` listener. We can't do this until we're actually firing listeners.

Reply:
Very fair and great feedback, I can make that change (don't remove until it's actually executed).